### ISSUE TITLE
Mutation hooks return non-memoized functions

### ISSUE THREAD

**Issue by rkfg** [marmelab/react-admin#8518]
I'm really unsure if it's a bug, a support question or a feature request but I don't see why it should work like that so I categorized it as a bug. The problem is that hooks like `useUpdate` or `useDelete` return a bare function, that is, a function that's not wrapped in `useCallback`. If I use it inside `useEffect` and pass it as a dependency, this effect fires on each render because `useUpdate` returns a new instance of the update function for no obvious reason.

**What you were expecting:**
Mutation hooks return the same memoized function unless their parameters change between renders.

**What happened instead:**
Mutation hooks return new function instances on each render.

**Steps to reproduce:**
Take the update function as `const [update] = useUpdate();`, then use it inside `useCallback` or `useEffect`. React linter would complain that it should be present in the dependency array and if you add it there the callback will update and effect will fire on each render. It's wasteful and might become a problem, for example, if you setup a timer in that effect and cancel it on cleanup. If the user keeps causing rerenders (that shouldn't affect the timer) it might be delayed a lot.

**Other information:**
The offending line is here: https://github.com/marmelab/react-admin/blob/bdff90eeb1e3e7012565c695d6cd7bf5d953d5a6/packages/ra-core/src/dataProvider/useUpdate.ts#L268 IMO, this function should be wrapped in `useCallback` but the list of dependencies might be long (I didn't try it). Same for `useUpdateMany`, `useDelete` and `useCreate`.

A quick and dirty way is to simply ignore the linter and not include the function as a dependency. I'm not sure if it causes any issues but so far it seems to work okay for me despite the warning. Not the cleanest way of course and it's not right: other side-effect hooks such as `useNotify` or `useRefresh` don't have this problem as they're properly wrapped in `useCallback`.

**Environment**

* React-admin version: 4.6.2
* Last version that did not exhibit the issue (if applicable): unsure
* React version: 18.1
* Browser: Brave 1.46.144 Chromium: 108.0.5359.128

**Comment by fzaninotto** [comment]
Hi, and thanks for your issue. You're right, the mutation callback isn't memoized and it can trigger useless re-renders. We need to memoize it.

But instead of adding a `useCallback` with a huge array of dependencies, we can use [the latest ref pattern](https://epicreact.dev/the-latest-ref-pattern-in-react/), i.e. something like:

```js
const useUpdate = (...) => {
   const mutation = useMutation(...);
   const updateRef = useRef(null);
   useLayoutEffect(() => {
      updateRef.current = async (callTimeResource, callTimeParams, updateOptions) => {
         /* the old update function */
       }
    });
    const update = useCallback(updateRef.current, []);
    return [update, mutation];
};
```

So the change should be pretty straightforward. Do you want to give it a try?

**Comment by rkfg** [comment]
Hmm, nice trick, thanks for sharing! I don't quite understand the purpose of `useLayoutEffect` here, can't we simply assign the function to `updateRef.current` in the main code? Without the dependency array it's called after every render anyway.

But this looks a small enough change so I'll take a look, sure!

**Comment by fzaninotto** [comment]
I was wondering the same question. It turns out React may stop in the middle of a render in concurrent mode, which is why [they're saying](https://beta.reactjs.org/apis/react/useRef):

> Do not write or read ref.current during rendering

You can use [the useEventCallback hook](https://github.com/marmelab/react-admin/blob/master/packages/ra-core/src/util/useEventCallback.ts) that is exported by `ra-core` - it's a shim for `useEvent` that does the layout effect on its own.

**Comment by fzaninotto** [comment]
Fixed by #8526
